### PR TITLE
Added test for multiple defaults and a new evaluate function

### PR DESCRIPTION
**Pull request by ourmoa**

### CODE REVIEW

File: puan/logic/plog/__init__.py [hunk 1]
```diff
@@ -558,25 +558,37 @@ def evaluate(self, interpretation: typing.List[puan.SolutionVariable]) -> bool:
...
-                                "evaluate", 
-                                interpretation=interpretation,
-                            ),
-                            self.compound_propositions
-                        )
-                    )
-                )
-            )-self.value
-        ) >= 0
+    def evaluate_propositions(self, dict: typing.Dict) -> dict:
```
Comment (znittzel) [inline puan/logic/plog/__init__.py:563]:
```suggestion
    def evaluate_propositions(self, dict: typing.Dict[str, int]) -> typing.Dict[str, int]:
```

File: puan/logic/plog/__init__.py [hunk 2]
```diff
@@ -558,25 +558,37 @@ def evaluate(self, interpretation: typing.List[puan.SolutionVariable]) -> bool:
                 ),
             )
         )
+        return self.evaluate_propositions(interpretation_map)[self.id] > 0
```
Comment (znittzel) [inline puan/logic/plog/__init__.py:561]:
```suggestion
        return self.evaluate_propositions(interpretation_map)[self.id] >= 0
```

File: puan/logic/plog/__init__.py [hunk 3]
```diff
@@ -558,25 +558,37 @@ def evaluate(self, interpretation: typing.List[puan.SolutionVariable]) -> bool:
...
+                >>> AtLeast(propositions=[puan.variable("x", dtype="int")], value=10).evaluate_propositions({"x": 10})
+                {'x': 10, 'VARa6aef82726db5033e4b25e6fec8b5770cf89fe44ff8336731eef2bfa9a8ab35f': 1}
+        """
+        def _get_variable_val(id, dict):
+            if not id in dict.keys():
+                dict[id] = 0
+            return dict
+        if not self.id in dict.keys():
+            # Calculate variable value and add to dict
+            dict[self.id] = 1*(sum(map( lambda x: x.evaluate_propositions(dict)[x.id]*self.sign if type(x)!=puan.variable else _get_variable_val(x.id, dict)[x.id]*self.sign, itertools.chain(self.atomic_propositions, self.compound_propositions))) -self.value >= 0)
```
Comment (znittzel) [inline puan/logic/plog/__init__.py:590]:
```suggestion
            dict[self.id] = 1*(
                sum(
                    itertools.chain(
                        map(
                            lambda x: x.evaluate_propositions(dict)[x.id]*self.sign,
                            self.compound_propositions
                        ),
                        map(
                            lambda x: _get_variable_val(x.id, dict)[x.id]*self.sign,
                            self.atomic_propositions
                        )
                    )
                ) -self.value >= 0
            )
```

File: puan/logic/plog/__init__.py [hunk 4]
```diff
@@ -558,25 +558,37 @@ def evaluate(self, interpretation: typing.List[puan.SolutionVariable]) -> bool:
...
-                            ),
-                            self.compound_propositions
-                        )
-                    )
-                )
-            )-self.value
-        ) >= 0
+    def evaluate_propositions(self, dict: typing.Dict) -> dict:
+        """
+            Evaluates propositions on this model given a dict with variabled and their values.
```
Comment (znittzel) [inline puan/logic/plog/__init__.py:565]:
```suggestion
            Evaluates propositions on this model given a dict with variables and their values.
```

File: puan/logic/plog/__init__.py [hunk 5]
```diff
@@ -558,25 +558,37 @@ def evaluate(self, interpretation: typing.List[puan.SolutionVariable]) -> bool:
...
+
+                >>> All(*"xy", variable="A").evaluate_propositions({"x": 1, "y":1})
+                {'x': 1, 'y': 1, 'A': 1}
+
+                >>> AtLeast(propositions=[puan.variable("x", dtype="int")], value=10).evaluate_propositions({"x":9})
+                {'x': 9, 'VARa6aef82726db5033e4b25e6fec8b5770cf89fe44ff8336731eef2bfa9a8ab35f': 0}
+
+                >>> AtLeast(propositions=[puan.variable("x", dtype="int")], value=10).evaluate_propositions({"x": 10})
+                {'x': 10, 'VARa6aef82726db5033e4b25e6fec8b5770cf89fe44ff8336731eef2bfa9a8ab35f': 1}
+        """
```
Comment (znittzel) [inline puan/logic/plog/__init__.py:583]:
```suggestion

            Returns
            -------
                out : typing.Dict[str, int]
        """
```

**Review by znittzel — CHANGES_REQUESTED** [review]

File: puan/logic/plog/__init__.py (hunk 2 above)
Comment (ourmoa) [reply puan/logic/plog/__init__.py:561]: No should be strict here. A zero = false, and 1 = True

File: puan/logic/plog/__init__.py (hunk 3 above)
Comment (ourmoa) [reply puan/logic/plog/__init__.py:590]: Nice!

File: tests/test_puan.py [hunk 6]
```diff
@@ -1644,6 +1644,23 @@ def test_not_when_single_str_or_variable():
...
+    full_config2 = model.evaluate_propositions(config2)
+    keys1 = numpy.array(list(full_config1.keys()))[:-1]
+    keys1.sort()
+    int_ndarray1 = numpy.array(list(map(lambda x: full_config1[x], keys1)))
+    keys2 = numpy.array(list(full_config2.keys()))[:-1]
+    keys2.sort()
+    int_ndarray2 = numpy.array(list(map(lambda x: full_config2[x], keys2)))
+    objective_function = puan.ndarray.ndint_compress(model.polyhedron.default_prio_vector, method="shadow")
+    assert objective_function.dot(int_ndarray1) > objective_function.dot(int_ndarray2)
+
```
[the remark this answers is not in the document]
Comment (ourmoa) [reply tests/test_puan.py:1663]: Row 1675 will always be true

File: puan/__init__.py [hunk 7]
```diff
@@ -45,6 +45,12 @@ class Bounds:
     lower: int = default_min_int
     upper: int = default_max_int
 
+    def __init__(self, lower: int, upper: int):
+        if lower > upper:
+            raise ValueError
```
Comment (znittzel) [inline puan/__init__.py:50]: Maybe just a description of what kind of value error?
```suggestion
            raise ValueError(f"upper bound must be higher than lower bound, got: ({lower}, {upper})")
```

File: puan/logic/plog/__init__.py [hunk 8]
```diff
@@ -523,72 +525,119 @@ def is_contradiction(self) -> bool:
...
+                            )
+                        )
+                _check_variable_in_bounds(prop.variable.id, 1*(val-prop.value >= 0), prop.variable.bounds)
+                interpretation[prop.variable.id] = 1*(
+                        val-prop.value >= 0
+                    )
+            return interpretation
+        interpretation_map = dict(map(lambda x: (x.id, x), filter(lambda x: type(x) == puan.variable, self.flatten())))
+        for x, y in interpretation.items():
+            if x in interpretation_map:
```
Comment (znittzel) [inline puan/logic/plog/__init__.py:638]:
```suggestion
        for x, y in filter(lambda x: x[0] in interpretation_map, interpretation.items()):
```

File: puan/ndarray/__init__.py [hunk 9]
```diff
@@ -1170,33 +1170,38 @@ def row_distribution(self, row_index: int) -> numpy.ndarray:
...
-                [3, 0],
-                [4, 1]
-            ])
-
-            2        |   
-            1  |     |     |
-            0 ---------------
-            0  1  2  3  4
+        
+            .. code-block::
```
Comment (znittzel) [inline puan/ndarray/__init__.py:1174]: Freeesshh!

**Review by znittzel — CHANGES_REQUESTED** [review]

**Review by znittzel — APPROVED** [review]

**Pull request merged**